data/src/main/java/org/apache/iceberg/data/GenericReader.java

```diff
@@ -0,0 +1,320 @@
+    if (eqDeletes.isEmpty()) {
+      return records;
+    }
+
+    Multimap<Set<Integer>, DeleteFile> filesByDeleteIds = Multimaps.newMultimap(Maps.newHashMap(), Lists::newArrayList);
+    for (DeleteFile delete : eqDeletes) {
+      filesByDeleteIds.put(Sets.newHashSet(delete.equalityFieldIds()), delete);
+    }
+
+    CloseableIterable<Record> filteredRecords = records;
+    for (Map.Entry<Set<Integer>, Collection<DeleteFile>> entry : filesByDeleteIds.asMap().entrySet()) {
+      Set<Integer> ids = entry.getKey();
+      Iterable<DeleteFile> deletes = entry.getValue();
+
+      Schema deleteSchema = TypeUtil.select(recordSchema, ids);
+
+      // a wrapper to translate from generic objects to internal representations
+      InternalRecordWrapper asStructLike = new InternalRecordWrapper(recordSchema.asStruct());
+
+      // a projection to select and reorder fields of the file schema to match the delete rows
+      StructProjection projectRow = StructProjection.create(recordSchema, deleteSchema);
+
+      Iterable<CloseableIterable<Record>> deleteRecords = Iterables.transform(deletes,
+          delete -> openDeletes(delete, dataFile, deleteSchema));
+      StructLikeSet deleteSet = Deletes.toEqualitySet(
+          // copy the delete records because they will be held in a set
+          CloseableIterable.transform(CloseableIterable.concat(deleteRecords), Record::copy),
+          deleteSchema.asStruct());
+
+      filteredRecords = Deletes.filter(filteredRecords,
+          record -> projectRow.wrap(asStructLike.wrap(record)), deleteSet);
+    }
+
+    return filteredRecords;
+  }
+
+  private CloseableIterable<Record> applyPosDeletes(CloseableIterable<Record> records, Schema recordSchema,
+                                                    CharSequence file, List<DeleteFile> posDeletes, DataFile dataFile) {
+    if (posDeletes.isEmpty()) {
+      return records;
+    }
+
+    Accessor<StructLike> posAccessor = recordSchema.accessorForField(MetadataColumns.ROW_POSITION.fieldId());
+    Function<Record, Long> posGetter = record -> (Long) posAccessor.get(record);
+    List<CloseableIterable<StructLike>> deletes = Lists.transform(posDeletes,
+        delete -> openPosDeletes(delete, dataFile));
+
+    // if there are fewer deletes than a reasonable number to keep in memory, use a set
+    if (posDeletes.stream().mapToLong(DeleteFile::recordCount).sum() < 100_000L) {
+      return Deletes.filter(records, posGetter, Deletes.toPositionSet(file, CloseableIterable.concat(deletes)));
+    }
+
+    return Deletes.streamingFilter(records, posGetter, Deletes.deletePositions(file, deletes));
+  }
+
+  private CloseableIterable<StructLike> openPosDeletes(DeleteFile file, DataFile dataFile) {
+    return openDeletes(file, dataFile, POS_DELETE_SCHEMA);
+  }
+
+  private <T> CloseableIterable<T> openDeletes(DeleteFile deleteFile, DataFile dataFile, Schema deleteSchema) {
```

Comment:
I saw the upper layer will need the `CloseableIterable<StructLike>` and `CloseableIteratable<Record>`,   How about marking the parameterized T as `T extends StructLike` here

Reply:
This should actually be `Record`. I'll fix it. Thanks for pointing this out.